kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/Utilities.java

```diff
@@ -0,0 +1,254 @@
+  }
+
+  // use reflection here to avoid requiring Hadoop as a dependency
+  private static Object loadHadoopConfig(IcebergSinkConfig config) {
+    Class<?> configClass =
+        DynClasses.builder().impl("org.apache.hadoop.hdfs.HdfsConfiguration").orNull().build();
+    if (configClass == null) {
+      configClass =
+          DynClasses.builder().impl("org.apache.hadoop.conf.Configuration").orNull().build();
+    }
+
+    if (configClass == null) {
+      LOG.info("Hadoop not found on classpath, not creating Hadoop config");
+      return null;
+    }
+
+    try {
+      Object result = configClass.getDeclaredConstructor().newInstance();
+      BoundMethod addResourceMethod =
+          DynMethods.builder("addResource").impl(configClass, URL.class).build(result);
+      BoundMethod setMethod =
+          DynMethods.builder("set").impl(configClass, String.class, String.class).build(result);
+
+      //  load any config files in the specified config directory
+      String hadoopConfDir = config.hadoopConfDir();
+      if (hadoopConfDir != null) {
+        HADOOP_CONF_FILES.forEach(
+            confFile -> {
+              Path path = Paths.get(hadoopConfDir, confFile);
+              if (Files.exists(path)) {
+                try {
+                  addResourceMethod.invoke(path.toUri().toURL());
+                } catch (IOException e) {
+                  LOG.warn("Error adding Hadoop resource {}, resource was not added", path, e);
+                }
+              }
+            });
+      }
+
+      // set any Hadoop properties specified in the sink config
+      config.hadoopProps().forEach(setMethod::invoke);
+
+      LOG.info("Hadoop config initialized: {}", configClass.getName());
+      return result;
+    } catch (InstantiationException
+        | IllegalAccessException
+        | NoSuchMethodException
+        | InvocationTargetException e) {
+      LOG.warn(
+          "Hadoop found on classpath but could not create config, proceeding without config", e);
+    }
+    return null;
+  }
+
+  public static Object extractFromRecordValue(Object recordValue, String fieldName) {
+    String[] fields = fieldName.split("\\.");
+    if (recordValue instanceof Struct) {
+      return getValueFromStruct((Struct) recordValue, fields, 0);
+    } else if (recordValue instanceof Map) {
+      return getValueFromMap((Map<?, ?>) recordValue, fields, 0);
+    } else {
+      throw new UnsupportedOperationException(
+          "Cannot extract value from type: " + recordValue.getClass().getName());
+    }
+  }
+
+  private static Object getValueFromStruct(Struct struct, String[] fields, int idx) {
+    Preconditions.checkArgument(idx < fields.length, "Invalid field index");
+    Object value = struct.get(fields[idx]);
+    if (value == null || idx == fields.length - 1) {
+      return value;
+    }
+
+    Preconditions.checkState(value instanceof Struct, "Expected a struct type");
+    return getValueFromStruct((Struct) value, fields, idx + 1);
```

Comment:
Why use recursion when this could use a loop instead?

Reply:
I changed this to a loop.